### PR TITLE
v8.1.0 Update 

### LINKED ISSUE #82: Repository Refactoring (proposal)
In order to avoid making a huge PR with generic changes, I'm going to write a tracking issue here that lists the changes I would like to propose,  in that way we will have a sort of changes tracking to isolate changes by 'type'. 

This issue supersedes #80, which involved a series of structural changes too delicate to be made all at once.

- [ ] #81 - this pr is used to add typescript as a dependency, use it to generate types and add type references inside files
- [ ] Move files from lib to src (including index.js)
- [ ] Remove deprecations (slice, substr), fix typos
- [ ] An esm module should not use cjs dependencies, check dependencies because perhaps not all of them are useful (nowadays at least).
- [ ] Update changelog, version and readme (require should be replaced with import)

### PR DESCRIPTION
**Pull request by erikyo**
this is the PR that contains all the changes listed in #82 

There are no dramatic changes, however, we may decide to move to 8.1.0 (and not 8.0.1) because of the modified type system, which did not require immediate action, but it would be preferable to suggest to remove the "definitively typed" types (if someone is using ts). For that reason a changelog file update is nice, hopefully someone is using `npm outdated`

**NOT ready to merge**

### CODE REVIEW

**Comment by smhg** [comment]
@erikyo thank you for your efforts!
If you want me to review these changes, please make it easy and separate things into separate PR's. Like i've been requesting.
Please put all cosmetic changes (imports, directory structure, typos,...) into one PR. I'm (by now) _very_ happy to quickly review that.
What then remains is only a limited set of changes related to TS which is harder for me to look at and the separation makes that review too _much_ easier.

**Comment by erikyo** [comment]
I understand your concerns, and I appreciate your feedback. Let me clarify my approach and workflow regarding the changes I've proposed.

The reason why you didn't see those changes in the master branch, is because I prefer to keep changes in a branch concurrent with the master branch. This has a simple reason... users that needs to download (or check the code) from the github repository should do so with the code from the latest release and not the code in development (which may not be stable). 
At this stage, i think it's better to keep these modifications on the `dev` branch, and this approach ensures that we can maintain a stable master branch until all planned modifications are finalized. 

At the moment I'm still missing a set of modifications (#89 which is now a draft) because I'd like @johnhooks to check it too, that guy is really good at this and his contribute would be really helpful

About typescript: The current changes primarily focus on stabilizing and updating the repository, rather than introducing TypeScript-related updates. The TypeScript compiler (tsc) is currently utilized solely for generating types and compiling JavaScript. This is why you'll notice `allowJs: true` in the [tsconfig.json](https://github.com/smhg/gettext-parser/pull/86/files#diff-b55cdbef4907b7045f32cc5360d48d262cca5f94062e353089f189f4460039e0R13). 

In summary, I intend to delay merging any changes to the master branch until I've completed the entire set of modifications I've outlined. This allows us to maintain a clear separation between development and stable code (if you want to try the development version, however, just checkout the development instead of the master and you will be on top of all the new updates). 

I hope this clarifies my approach, but I welcome any further discussion or suggestions you may have.

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import { Transform } from "readable-stream";
+
+export declare module 'encoding' {
```
Comment (johnhooks) [inline src/index.d.ts:3]: @erikyo this shouldn't be in the types exported by `gettext-parser` it should be added to a local declaration file, and included in the `"compilerOptions"."types"` of the `tsconfig.json` otherwise the declaration of another module is being exported as part of the `gettext-parser` package, which I don't believe is the intent.

File: package.json [hunk 2]
```diff
@@ -18,28 +18,36 @@
...
     "test-generate-mo": "msgfmt test/fixtures/latin13.po -o test/fixtures/latin13.mo & msgfmt test/fixtures/utf8.po -o test/fixtures/utf8.mo & msgfmt test/fixtures/obsolete.po -o test/fixtures/obsolete.mo",
     "test": "mocha",
     "preversion": "npm run lint && npm test",
-    "postversion": "git push && git push --tags"
+    "postversion": "git push && git push --tags",
+    "prepublishOnly": "npm i && tsc && npm run lint && npm run test"
   },
-  "main": "./index.js",
+  "main": "./lib/index.js",
+  "types": "./lib/index.d.ts",
```
Comment (johnhooks) [inline package.json:29]: Why manually write the declaration and not use what is generated by `tsc`?

**Review by johnhooks — CHANGES_REQUESTED** [review]
@erikyo the details are in the comments, though my biggest question: why manually create a type declaration, rather than use one generated by `tsc`? It looks like you have it configured in the `tsconfig.json` for it to generate a declaration file.

**Comment by johnhooks** [comment]
I've read through the previous comments for this PR. I agree with @smhg, it would be best to create separate PRs for the following :
- Moving `lib` to `src`.
- Updating of dependencies
- Refactoring the usage of old APIs.
- Converting away from regex for finding comments and quotes.
- Adding the initial generation of type declarations.

  For this issue, I don't see a huge benefit in implementing the types in JSDocs, if there is support for converting to TypeScript.

File: package.json (hunk 2 above)
Comment (erikyo) [reply package.json:29]: You must consider that we must have continuity with past types, without changing the name. For this we will have to keep a declaration file in which we have these named 

The declaration file ["src/index.d.ts" is used by tsc](https://github.com/smhg/gettext-parser/blob/955eaa25aa0cb3e117fdf0742d66bcc6afeb3601/src/index.d.ts#L12-L37) to generate a new one that is outputted to the `lib/` directory (the build folder, BTW everything you see inside "lib" is now generated).

Anyhow the fix for the types is not yet merged at this stage in this branch you should look at #89 for that (actually is a draft)

File: src/index.d.ts (hunk 1 above)
Comment (erikyo) [reply src/index.d.ts:3]: ah right, but it's really bad to have to make a file just for this dependency which is not typed. Tn the pr #89 where the types were fixed I export it like [in this way](https://github.com/smhg/gettext-parser/blob/42a969325c052b9b1927291b4d6a15172298e21f/src/types.d.ts)... I will try to find a way to avoid it if possible, if you have a proposal please tell me

**Comment by erikyo** [comment]
>  why manually create a type declaration, rather than use one generated by tsc?

Because there are 400 errors in types and the generated output isn't what we expect at the moment. What you say would be ideal but if you try to generate the file completely the result now is not what you expect. 

To fix it completely requires either very radical changes (or as I am doing several pr). This is somewhat the reason for the dev branch (Release Branch Strategy). The reason was because the changes required are too many to 'fix everything' but at the same time an intermediate version cannot be published on the master.

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    compile: (table: GetTextTranslations, options?: parserOptions) => Buffer;
+    createParseStream: (options?: parserOptions, transformOptions?: import('readable-stream').TransformOptions) => Transform;
+}
+
+export interface mo {
+    parse: (buffer: Buffer | string, defaultCharset?: string) => GetTextTranslations;
+    compile: (table: GetTextTranslations, options?: parserOptions) => Buffer;
+}
+
+export * from "./@types";
```
[the remark this answers is not in the document]
Comment (erikyo) [reply src/index.d.ts:50]: To me what is is not 100% correct of that, is to export the types related to the functions provided by gettext-parser in i.e. this file [this](https://github.com/smhg/gettext-parser/blob/42a969325c052b9b1927291b4d6a15172298e21f/src/types.d.ts) doing this https://github.com/smhg/gettext-parser/blob/42a969325c052b9b1927291b4d6a15172298e21f/src/types.d.ts#L37-L46

but this is required i guess until we decide to switch completely to typescript because the jsdoc didn't provide the required output. Furthermore these are exactly like the ones that are now on the [strictly typed repo](https://www.npmjs.com/package/@types/gettext-parser?activeTab=code), so, in order to avoid any possible glitches for users I had to take them and restore (approximately) as they were before

File: src/index.d.ts (hunk 1 above)
Comment (erikyo) [reply src/index.d.ts:3]: Just checked and seem that the `src/index.d.ts` file is not "copied" into `lib` when building the module... this partially solves the issue you mentioned but, on the other hand i need to find a way to export all the required definitions (i.e. `GetTextTranslation` or `GetTextComment`)

File: src/index.d.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (erikyo) [reply src/index.d.ts:50]: fixed with https://github.com/smhg/gettext-parser/pull/89/commits/cbe6e8f05e6455ea79023b12af39d20077b6c1d2

Since having a `.d.ts` declarations file would have complicated things I moved the definitions into a separate `.js` file. In that way we get the same result because tsc will generates a nearly identical definitions file them during compilation

File: package.json (hunk 2 above)
Comment (erikyo) [reply package.json:29]: fixed in https://github.com/smhg/gettext-parser/commit/cbe6e8f05e6455ea79023b12af39d20077b6c1d2

**Comment by johnhooks** [comment]
In my opinion, types aren't currently published by the package,so it doesn't make sense to copy the Definitely Typed ones. Those are established and what people are used to. It would be best to wait and publish the actual types when they are ready.

I wouldn't copy the types from Definitely Typed, but enable types with very low strictness, and start improving the situation in increments. The types can be incomplete and unpublished (by unpublished, I mean not adding the `"types"` property to the `package.json` until they are ready). This would not require a full overhaul of the package in one large step. A build step can still be used to convert `ts` to `js` files. Flip them all to `ts` as they are and accept they will not pass a strict type check, or any type check, but then it can be worked at in digestible chunks.

**Comment by erikyo** [comment]
Sorry John, maybe I didn't provide enough context. The types from Definitely Typed are for version 4.0.4 and not for the current 8.0.0, and they are indeed very generic, sometimes incorrect (like in the case of comments where they are all marked as required when they're not) and they only cover the exported functions in the index. When I mentioned using the same types, I was referring to the naming convention chosen for the translation block, whether it's "GetTextTranslation" or "GetTextTranslations".

So, #89 isn't a copy of Definitely Typed, but it's obviously very similar since the functions and types are the same (as I mentioned earlier, even the names). However, they are correct, which isn't the case for Definitely Typed.

I'm taking steps forward: in #81, I started generating types (with some issues because index.js and index.d.ts were in the root), then in #85, I moved all the files into src (which resolved the problem), etc. Making it strictly typed requires changing or adding a lot of code, unfortunately not all JSDoc comments were correct, and this results in many lines to be changed, and consequently, if not one giant PR, then many small PRs (which is what I'm attempting to do).

**Comment by johnhooks** [comment]
Thanks @erikyo for the explanation, I'd would still lean toward waiting to publish them until they are "correct" and generated by properly typed code. IMHO

@smhg Could we please enable discussions for this repository? I think that could be a better place to hash out the details of git strategy, rather than in a PR or Issue

**Comment by erikyo** [comment]
@johnhooks After a bit of tweaking yesterday I managed to reduce the errors to 280, so I don't think the result is that far off. 

you should try the #89 using `"checkJs": false` [in tsconfig](https://github.com/smhg/gettext-parser/blob/7f94d007e3a6d3f8b05180d6d71d9337fcdadd8b/tsconfig.json#L16C4-L16C22) you would have a clearer situation 😉

**Comment by smhg** [comment]
@johnhooks  I've started one discussion topic. Hopefully that does the trick. Please tell me if this isn't set up properly.

File: src/mocompiler.js [hunk 4]
```diff
@@ -183,9 +222,9 @@ Compiler.prototype._calculateSize = function (list) {
 /**
  * Generates the binary MO object from the translation list
  *
- * @param {Array} list translation list
- * @param {Object} size Byte size information
- * @return {Buffer} Compiled MO object
+ * @param {TranslationBuffers[]} list translation list
+ *  @param {Size} size Byte size information
+ *  @return {Buffer} Compiled MO object
```
Comment (johnhooks) [inline src/mocompiler.js:227]: We should remove these spaces... also would be nice to add a JSDoc eslint rules.

**Review by johnhooks — COMMENTED** [review]
This is looking really good. I found just a few extras in a JSDocs.

File: src/mocompiler.js (hunk 4 above)
Comment (erikyo) [reply src/mocompiler.js:227]: yep you are right, it seems that jsdocs aren't formatted at all. 

What do you suggest to fix that issue?

Comment (johnhooks) [reply src/mocompiler.js:227]: @erikyo I've got a branch that is almost done, in which I manually format them. I looked into using `eslint-plugin-jsdoc` though I was having some trouble because `eslint-config-standard` hasn't be updated (by it's maintainers) and I would need to figure out what version of `eslint-plugin-jsdoc` should be added as a dev dependency. If there is support to add the JSDoc plugin, I can figure it out, but didn't want to commit much time to it, if there wasn't support.
